File: frontend/src/types/budget.ts
export interface Budget {
  id: string;
  user_id: string;
  category_id?: string | null; // null = overall monthly budget or daily limit
  category_name?: string | null;
  amount: number;
  period: 'monthly' | 'daily';
  created_at: string;
  updated_at: string;
}

export interface BudgetStatus {
  total_budget: number;
  total_spent: number;
  remaining: number;
  percentage_used: number;
  status: 'on_track' | 'near_limit' | 'over_budget';
}

export interface DailyLimitStatus {
  daily_limit: number;
  today_spent: number;
  remaining: number;
  percentage_used: number;
  status: 'on_track' | 'near_limit' | 'over_budget';
}

File: backend/app/schemas/budget.py
import uuid
from datetime import datetime
from decimal import Decimal
from typing import Optional, Literal
from pydantic import BaseModel, Field, ConfigDict

class BudgetBase(BaseModel):
    category_id: Optional[uuid.UUID] = Field(None, description="Category ID or null for overall budget")
    amount: Decimal = Field(..., gt=0, decimal_places=2, description="Monthly budget limit in INR (₹)")
    period: Literal["monthly"] = "monthly"

class BudgetCreate(BudgetBase):
    pass

class BudgetUpdate(BaseModel):
    amount: Decimal = Field(..., gt=0, decimal_places=2)

class BudgetResponse(BudgetBase):
    id: uuid.UUID
    user_id: uuid.UUID
    category_name: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)

class BudgetStatusResponse(BaseModel):
    total_budget: Decimal
    total_spent: Decimal
    remaining: Decimal
    percentage_used: float
    status: Literal["on_track", "near_limit", "over_budget"]
